drift.py

```diff
@@ -1,146 +1,90 @@
+    version:str
     commit:str
     tags:list
     timestamp:str
-    spec:str
-    version:str
-
-    def __init__(self):
-        self.abstract = ""
-        self.commit = ""
-        self.tags = []
-        self.timestamp = ""
-        self.spec = ""
-        self.version = ""
-
 
 class DriftAnalysis(Helicase):
-    specs = []
-    last = {}
-    def __init__(self, specs):
-        self.specs = specs
-        for spec in specs:
-            self.last[spec] = ""
+    def __init__(self, specs=None):
+        self.last = {}
+        self.specs = specs or []
 
     def analyze(self, commit):
-        for spec in self.specs:
-            out = spack("spec --json " + spec)
-            verToken = spec.find("@")
-            if verToken < 0: verToken = len(spec)
-            name = spec[:verToken]
-            result = Result()
-            result.commit = commit.hash
-            result.abstract = name
-            result.timestamp = str(commit.author_date)
-            if len(spec[verToken:]) > 0:
-                result.version = spec[verToken+1:]
-            if out.returncode == SUCCESS:
-                concrete_spec = json.loads(out.stdout)["spec"]
-                # In the case that the concretized spec hash changes
-                # we want to record the commit where this happens.
-                if self.last[spec] != "" and self.last[spec][0][name]["full_hash"] != concrete_spec[0][name]["full_hash"]:
-                    # Check if the package version is different.
-                    if self.last[spec][0][name]["version"] != concrete_spec[0][name]["version"]:
-                        result.tags += ["pkg-updated"]
-                    # +---------------------+
-                    # |     Dependencies    |
-                    # +---------------------+
-                    result.tags += set(self.tag_deps(concrete_spec, spec, name))
-                    # +---------------------+
-                    # |      Variants       |
-                    # +---------------------+
-                    result.tags += set(self.tag_variants(concrete_spec, spec, name))
-                    # Send Inflection Point
-                    send(result)
-                # Save concrete spec hash as last hash.
-                self.last[spec] = concrete_spec
-            else:
-                # If the spec doesn't concretize properly we also want
-                # to record the commit at which this occurred.
-                result.tags += ["concretizaton-failed"]
+        for abstract_spec in self.specs:
+            # Separate out spec version and name.
+            name = abstract_spec
+            version = ""
+            if "@" in abstract_spec:
+                name, version = abstract_spec.split("@", 1)
+            # Check that version exists in package.py file.
+            out = run(f"spack versions --safe-only {abstract_spec}")
+            # Don't attempt to concretize if the version doesn't yet exist.
+            if version in out.stdout:
```

Comment:
This will succeed erroenously if, e.g., your version is `1.0` and stdout contains `5.1.0`

Reply:
Oh good point! But why not just split the string on whitespace using `split()`?

Reply:
```python
>>> "   this   has   lots  of\t\tspaces\t and tabs ".split()
['this', 'has', 'lots', 'of', 'spaces', 'and', 'tabs']
```

Apparently that works fine -- if it gets rid of all the whitespace then great -- I'm used to having to do `r"\s+"` to get rid of more whitespace but it looks like I didn't realize that `split()` did that by default.